### ISSUE TITLE
Variables can't be declared using both 'var' and a type name.

### ISSUE THREAD

**Issue by samuelstroschein** [dart-lang/language#1551]
Hi everyone, 

Declaring a variable with a type name is not possible. However, declaring a final variable with a type name is possible. Not only do I find that unintuitive, but additionally, I find using the var keyword more explicit and thus more readable than just declaring a "variable" without the var keyword. I observed in other people's code that writing var/final instead of the type generally seems to be preferred. 

The following is a very short example:

```Dart
String x = "this works";
var String y = "this does not work";
final String z = "this works";
```

Line 2, 3 immediately signal a variable assignation, while line 1 is less explicit. 

By purely looking at the example code, I would even make the argument that line 1 should not work. 

~~**Is there a technical reason behind the differentiation?**  
If not, why should line 2 be illegal? Line 2 is more consistent with TypeScript, Swift, and Kotlin.~~

### Update
My issue revolves around enabling/allowing the declaration of type annotated vars, as shown in line 2. It seems like this should just work. So let me rephrase my question:

**Are there arguments against allowing the declaration of type annotated variables as in line 2 (optional to the default declaration as shown in line 1)**?

**Comment by passsy** [comment]
This is often a problem for beginners when refactoring. Adding a type afterwards - for clarity - works differently for `final` and `var` variables:

```dart
// Type can be added after final
final result = await api.loadUser();
final UserResponse result = await api.loadUser();
```

```dart
// Type has to replace var
var result = await api.loadUser();
UserResponse result = await api.loadUser();
```

```dart
// illegal
var UserResponse result = await api.loadUser();
```

**Comment by samuelstroschein** [comment]
@passsy Thanks for the reply. That's more or less what I wrote. Do you know if there is a reason why var <T> name is illegal?

**Comment by lrhn** [comment]
Writing `String foo;` or `final String foo;` is the default. That's how you declare a typed variable.
Since Dart also allowed untyped code (especially in Dart 1), you were allowed to write `var foo;` as a shorthand for `dynamic foo;`, in case you didn't want to bother writing the type, because the Dart syntax, with its C ancestry, needs *some* prefix in order to declare a variable.

So, writing `var` is what you do *instead* of writing a type. It was intended as a shorthand. Writing both `var` and a type would miss the point of the shorthand, and it would provide no advantage over just the type, so it was not allowed.

**Comment by samuelstroschein** [comment]
@lrhn I assumed there were some historical reasons. But besides historical reasons, why should this be illegal? It's not consistent with final and I assume the shorthand (not using the keyword var) is less readable, especially if coming from TypeScript, Kotlin, Swift.

**Comment by awkj** [comment]
I think this is a compromise, Dart when the first generation of designers chose to look like C syntax declarations, when faced with automatic derivation, only this syntax can be used, so Typesctipt Kotlin Swift they have chosen to posttype, this syntax for type derivation looks more reasonable

**Comment by tatumizer** [comment]
There's an (unexpected?) connection between "var", "fn"/"fun" on one side, and optional semicolons on another.
The languages that have chosen to make semicolons optional have to pay a price - they **need** var, fn and friends, and make other syntax compromises, otherwise the parsing would be difficult/impossible. You cited Typescript, Kotlin, Swift - what do they have in common? Optional semicolons. Dart doesn't have to pay that price.

**Comment by samuelstroschein** [comment]
@tatumizer That's indeed an interesting connection. I get that the argument could be made that not having to declare var is a benefit as it's shorter but Dart/Flutter tries to be as explicit as possible e.g. `column` instead of `col` etc. For the sake of explicit declaration of a variable, I would think that one should be able to declare a type annotated var, just like the declaration of a type annotated final variable works. But the sentiment in this thread seems to disagree.

**Comment by tatumizer** [comment]
The use of "var" keyword in dart **today** is closely related to type inference. 
```
var x=0, y="hello";
```
If you imagine for a second that dart doesn't have type inference - then you would probably never use "var" at all.
In the past, dart was not strongly typed, and "var x" without initializer was used as a (shorter) synonym to "dynamic x".
This meaning is still preserved today, but mostly (as I speculate) for backward-compatibility reasons: "dynamic" vars are not very common in dart today (at least not as explicit declarations), so there's no justification for having a short notation - you could as well write "dynamic x" in the rare cases where you wanted that meaning.

The bottom line: if you want type inference, use "var". Otherwise, don't use "var" - declare your type explicitly.
Saying `var String x='hello'` would go against this practice. What's the point? If you say `var x='hello'`, it will infer "String" anyway - so it's "String" that is redundant, not "var".

With type inference, there's another interesting property of "var": you can declare variables of different types in the same statement: `var x=0, y='hello';`. Even if the compiler allowed typed var declarations for single variables, it would still reject this:
`var x=0, String y='hello';` (by analogy with final:  `final x=0, String y='hello';` is an error).

**Comment by samuelstroschein** [comment]
Thanks again for the in-depth answer. I feel like I have to clarify again that I know how to write Dart code. Some days ago, I just stumbled upon this error and thought this should not be an error. I do find explicitly using the var keyword more readable than just using a type to declare a variable. Likely syntax highlighting plays a role as well (blue for variable assignation vs green for type declaration, which is all over the code) but I still don't see any reason why **optionally** declaring a type annotated var should not be possible as it is not consistent with final, less explicit thus minimally less readable and less "similar" to Kotlin, TypeScript, and Swift. 

But again, seems like I am alone with that view.

**Comment by samuelstroschein** [comment]
You raise a valid point that I was not considering. 

I guess the issue boils down to more similarity with TS,Kotlin, Swift, Rust etc. which is related to the infamous issue #69. I will close this issue. 

Thanks @tatumizer for the discussion.
